### PR TITLE
Monitor batch write

### PR DESCRIPTION
**Pull request by lifenod**

### CODE REVIEW

**Comment by lifenod** [comment]
@bodhi 

I change to channel implementation.

> What I was going to say was: Here's an (older) reference that says 5-10k events per request is a good amount: https://community.influxdata.com/t/what-is-the-highest-performance-method-of-getting-data-in-out-of-influxdb/464

I found this in the influxdb doc: https://docs.influxdata.com/influxdb/v1.7/concepts/glossary#batch

> InfluxData recommends batch sizes of 5,000-10,000 points, although different use cases may be better served by significantly smaller or larger batches.

File: monitoring/influxdb.go [hunk 1]
```diff
@@ -105,52 +210,141 @@ type influxdbMonitor struct {
...
+func (ce *cacheEvent) Reset() {
+	// This if is rarely true, so add it.
+	if ce.currentCacheEvents != ce.cacheEvents {
+		ce.currentCacheEvents = ce.cacheEvents
+	}
+}
+
+func (im influxdbMonitor) batchWriteTicker() {
+	var points []*influxdb.Point
+	t := time.NewTicker(im.batchWriteInterval)
```
Comment (bodhi) [inline monitoring/influxdb.go]: `time.Tick` gives you the channel directly. You don't call `time.Ticker.Stop` anyway so you only need the channel.

---

I don't think this is a big problem, but you should consider using `time.After` instead.

`time.Ticker` doesn't work as a "timeout", and using `time.Ticker` will result in this sometimes:

1. read point from channel at time T, reach maximum point count, send points to InfluxDB

2. read point from channel at time T + 10ms, add point to buffer

3. read from ticker channel T + 20ms (or something similar), send a HTTP request with one point

=> "Send data to InfluxDB when we have `MaxCacheEvents` *or* every `batchWriteInterval` seconds

Vs:

1. read point from channel at time T, reach maximum point count, send points to InfluxDB, reset timeout

2. read point from channel at time T + 10ms, add point to buffer

3. can't read from timeout channel until `T + im.batchWriteInterval`

=> "Send data to InfluxDB when we have `MaxCacheEvents` *or* we haven't sent data since `batchWriteInterval` seconds ago"

So if you have a lot of data to send (> ~5000 points/minute), the timeout will never need to trigger.

File: monitoring/influxdb.go [hunk 2]
```diff
@@ -92,11 +195,13 @@ func NewInfluxdbMonitor(config InfluxMonitorConfig, logger log.Logger) (Monitor,
 		}
 	}()
 
-	logger.Info().Log(
-		"msg", fmt.Sprintf("influxdb instrumentation writing to %s://%s@%s/%s", u.Scheme, username, u.Host, monitor.database),
+	go monitor.batchWriteTicker()
```
Comment (bodhi) [inline monitoring/influxdb.go]: What happens if `monitor.batchWriteTicker` panics?

File: monitoring/influxdb.go [hunk 3]
```diff
@@ -105,52 +210,141 @@ type influxdbMonitor struct {
...
+	pointChan          chan *influxdb.Point
+	batchWriteInterval time.Duration
+	cacheEvent         *cacheEvent
 }
 
-// InsertRecord part of monitor.Monitor.
-func (im influxdbMonitor) InsertRecord(measurement string, value interface{}, tags map[string]string, fields map[string]interface{}, at time.Time) {
-	if fields == nil {
-		fields = map[string]interface{}{}
+type cacheEvent struct {
```
Comment (bodhi) [inline monitoring/influxdb.go]: I'm having trouble understanding the purpose of this struct. Can you please add some comments to describe how it is used?

File: monitoring/influxdb.go [hunk 4]
```diff
@@ -105,52 +210,141 @@ type influxdbMonitor struct {
...
 	if err != nil {
-		l.Error().Log(
+		_ = im.logger.Error().Log(
+			"database", im.database,
 			"err", err,
 			"during", "influxdb.NewBatchPoints",
-			"msg", fmt.Sprintf("Error initializing batch points for %s: %v", measurement, err),
+			"msg", fmt.Sprintf("NewBatchPoints failed: %v", err),
 		)
+		return nil
```
Comment (bodhi) [inline monitoring/influxdb.go]: This should return `err`, not `nil`?

File: monitoring/influxdb.go [hunk 5]
```diff
@@ -105,52 +210,141 @@ type influxdbMonitor struct {
...
-		l.Error().Log(
+		_ = im.logger.Error().Log(
+			"database", im.database,
 			"err", err,
-			"during", "influxdb.NewPoint",
-			"msg", fmt.Sprintf("Error initializing a point for %s: %v", measurement, err),
+			"during", "influxdb.client.Write",
+			"msg", fmt.Sprintf("influxdb client write points failed: %v", err),
 		)
+		return errors.Errorf("influxdb client write points failed: %v", err)
```
Comment (bodhi) [inline monitoring/influxdb.go]: Please use `errors.Wrap`.

**Comment by lifenod** [comment]
> I find it hard to follow how you've implemented the maximum buffer size, and what the use of `type cacheEvent` is. As I mentioned, please add some comments (or perhaps the design can be simplified?)

`type cacheEvent struct {`

> I'm having trouble understanding the purpose of this struct. Can you please add some comments to describe how it is used?

Let me explain the logic of `CacheEvents` and `MaxCacheEvents` first.

- If cached points count reach `CacheEvents` then trigger batch writing.
- If batch write fails, then we need `MaxCacheEvents` as the maximum cache limit.
  If batch write fails, then we increase the cache size (CurrentCacheEvents = CurrentCacheEvents + CacheEvents).
  For example `CacheEvents = 5000, MaxCacheEvents = 15000`:
    1. set `CurrentCacheEvents = CacheEvents = 5000`
    2. cached points count reach 5000
    3. trigger batch writing, but write fails
    4. we increase the `CurrentCacheEvents`, `CurrentCacheEvents` will be `5000 + 5000 = 10000`
    5. cached points count reach 10000
    6. trigger batch writing, but write fails
    7. we increase the cache size, `CurrentCacheEvents` will be `10000 + 5000 = 15000`
    8. cached points count reach 15000
    9. trigger batch writing, but write fails
    10. we increase the cache size, `CurrentCacheEvents` will be `MaxCacheEvents = 15000`, because `15000 + 5000 = 20000 > MaxCacheEvents`
    11. at this time, cached points count (15000) reach `MaxCacheEvents`, we will clean up the all cache points
    12. Once write successfully, we will reset `CurrentCacheEvents = CacheEvents = 5000`

Use `cacheEvent` to manage `CurrentCacheEvents`, `CacheEvents` and `MaxCacheEvents`, so `cacheEvent` has `CurrentCacheEvents`, `IncreaseCacheEvents` and `Reset` functions.

**Comment by bodhi** [comment]
Small English issue: "cache" isn't the correct word for this concept, "buffer" is. What you are doing is "buffering writes".

---

~Why can't you just use `len(points)`?~

Ah, because then *every point received* will trigger a new attempt to send the buffer to InfluxDB once `len(points) > CacheEvents`?

---

```
type cacheEvent struct {
	cacheEvents        int
	maxCacheEvents     int
	currentCacheEvents int
}
```

I think this would make it clearer:

1. Use cacheEvents and maxCacheEvents directly from the config, don't store them in `cacheEvent`. This helps to show that these values are *read-only*.

2. Rename `cacheEvents` to `nextWriteCount` or something like that, to mean "when the buffer gets to this size, we will attempt to write the buffer to InfluxDB".

---

You're using it in a flow like this?

flow

You only call `*cacheEvent.Reset` after a successful write. I think this means (assuming limit 5k, max 15k):

1. Receive 5k points
2. Hit limit. Attempt to write to InfluxDB
3. Write fails
4. Increase limit to 10k
5. Receive 5k points (buffer now has 10k points)
6. Hit limit. Attempt to write to InfluxDB
7. Write fails
8. Increase limit to 15k
9. Receive 5k points (buffer now has 15k points)
10. Hit limit. Attempt to write to InfluxDB
11. Write fails
12. buffer is at max length, drop all points
13. Receive 5k points
14. *Do nothing, since limit is still at 15k.* <= we won't attempt another write until we get 15k points...

You mentioned something like this in your comment:

> Once write successfully, we will reset `CurrentCacheEvents = CacheEvents = 5000`

But it seems like strange behaviour to leave `CurrentCacheEvents` at 15k after the buffer is dropped.

**Comment by lifenod** [comment]
> Ah, because then every point received will trigger a new attempt to send the buffer to InfluxDB once len(points) > CacheEvents?

Yes, if we keep points when writing fails, then len(points) > CacheEvents always is true.

> You're using it in a flow like this?

Yes.

> But it seems like strange behaviour to leave CurrentCacheEvents at 15k after the buffer is dropped.
> 14. Do nothing, since limit is still at 15k. <= we won't attempt another write until we get 15k points...

At this time, InfluxDB is still broken, so we keep CurrentCacheEvents = MaxCacheEvents for not send buffer to InfluxDB many times, but we also have the ticker to trigger buffer to InfluxDB. In other words, the batch write is MaxBuffer status at this time.

File: monitoring/influxdb.go [hunk 6]
```diff
@@ -42,35 +76,104 @@ func NewInfluxdbMonitor(config InfluxMonitorConfig, logger log.Logger) (Monitor,
...
+		if interval != "" {
+			second, err := strconv.Atoi(interval)
+			if err != nil {
+				return nil, errors.Wrapf(err, "influxdb config parameter %s format error", batchWriteSecondIntervalParamName)
+			}
+
+			batchWriteInterval = time.Duration(second) * time.Second
+		}
+	}
+	if batchWriteInterval == 0 {
```
[the remark this answers is not in the document]
Comment (lifenod) [reply monitoring/influxdb.go]: Fixed in https://github.com/theplant/appkit/blob/70c4f66134115cb94f1b703e0807a09a305c5fb5/monitoring/influxdb.go#L98.

File: monitoring/influxdb.go [hunk 7]
```diff
@@ -42,35 +76,104 @@ func NewInfluxdbMonitor(config InfluxMonitorConfig, logger log.Logger) (Monitor,
...
 	}
 
-	client, err := influxdb.NewHTTPClient(httpConfig)
+	values := u.Query()
+
+	var batchWriteInterval time.Duration
+	{
+		interval := values.Get(batchWriteSecondIntervalParamName)
+		if interval != "" {
+			second, err := strconv.Atoi(interval)
```
[the remark this answers is not in the document]
Comment (lifenod) [reply monitoring/influxdb.go]: Fixed in https://github.com/theplant/appkit/blob/70c4f66134115cb94f1b703e0807a09a305c5fb5/monitoring/influxdb.go#L91.

File: monitoring/influxdb.go (hunk 1 above)
Comment (lifenod) [reply monitoring/influxdb.go]: Good point.

Changed `time.Ticker` to `time.After` in https://github.com/theplant/appkit/pull/25/commits/18f119747a2cb76ede9d7f36579a3b7e9fcd5758.

File: monitoring/influxdb.go (hunk 2 above)
Comment (lifenod) [reply monitoring/influxdb.go]: Added `recover` in https://github.com/theplant/appkit/pull/25/commits/6edac7082280b7b75f55382bad87371d22b5a7e4.

File: monitoring/influxdb.go (hunk 3 above)
Comment (lifenod) [reply monitoring/influxdb.go]: Removed `cacheEvent` in https://github.com/theplant/appkit/pull/25/commits/af37ccc7faaceb0087dc23b229d3911c1ad74011.

File: monitoring/influxdb.go (hunk 4 above)
Comment (lifenod) [reply monitoring/influxdb.go]: Fixed in https://github.com/theplant/appkit/blob/70c4f66134115cb94f1b703e0807a09a305c5fb5/monitoring/influxdb.go#L302.

File: monitoring/influxdb.go (hunk 5 above)
Comment (lifenod) [reply monitoring/influxdb.go]: Fixed in https://github.com/theplant/appkit/blob/70c4f66134115cb94f1b703e0807a09a305c5fb5/monitoring/influxdb.go#L315.

**Comment by lifenod** [comment]
> For each request, please also include an extra data point into a separate InfluxDB measurement that logs the number of points in the request. Something like this:
> 
>    ```
>    [req1, req2, ..., reqN, { "measurement": "influxdb-queue-length", value: N }]
>    ```
> 
>    Then we can also monitor the behaviour of the buffer/queue itself.

Added in https://github.com/theplant/appkit/pull/25/commits/70c4f66134115cb94f1b703e0807a09a305c5fb5.

> Rename cacheEvents to nextWriteCount or something like that, to mean "when the buffer gets to this size, we will attempt to write the buffer to InfluxDB".

> Small English issue: "cache" isn't the correct word for this concept, "buffer" is. What you are doing is "buffering writes".

Renamed in https://github.com/theplant/appkit/pull/25/commits/af37ccc7faaceb0087dc23b229d3911c1ad74011. Renamed all `cacheEvents` to `bufferSize`.
